allennlp/nn/samplers/samplers.py

```diff
@@ -0,0 +1,125 @@
+import torch
+
+from allennlp.nn.samplers.sampler import Sampler
+from allennlp.nn.util import min_value_of_dtype
+
+
+@Sampler.register("multinomial")
+class MultinomialSampler(Sampler):
+    """
+    Represents a sampler to choose calues from a multinomial distribution
+
+    Registered as a `Sampler` with name "multinomial".
+    """
+
+    def __init__(self, temperature: float = 1.0, filter_val: float = -float("inf")) -> None:
+        self.temperature = temperature
+        self.filter_val = min_value_of_dtype(torch.float)
+
+    def __call__(
+        self, logits: torch.Tensor, num_samples: int = 1, with_replacement: bool = True
+    ) -> torch.Tensor:
+        probabilities = torch.nn.functional.softmax(logits, dim=-1)
+        selected_indices = torch.multinomial(
+            probabilities, num_samples, replacement=with_replacement
+        )
+
+        return (torch.gather(logits, 1, selected_indices), selected_indices)
+
+
+@Sampler.register("TopK")
+class TopKSampler(Sampler):
+    """
+    Represents a Sampler which redistributes the probability mass function among
+    the top `k` choices then selects from that subset
+    `logits` is a tensor of log-probabilities to be selected from.
+    `k` is the number of highest-probability options that the returned choice will be selected from
+    `temperature` modules the probabilitis of the selected tokens. A `temperature` below 1.0 produces a
+    sharper probability distribution and a `temperature` above 1.0 produces a flatter probability
+    distribution.
+
+    Registered as a `Sampler` with name "TopK".
```

Comment:
By convention we generally use all lowercase with hyphens for registered names:

```suggestion
@Sampler.register("top-k")
class TopKSampler(Sampler):
    """
    Represents a Sampler which redistributes the probability mass function among
    the top `k` choices then selects from that subset
    `logits` is a tensor of log-probabilities to be selected from.
    `k` is the number of highest-probability options that the returned choice will be selected from
    `temperature` modules the probabilitis of the selected tokens. A `temperature` below 1.0 produces a
    sharper probability distribution and a `temperature` above 1.0 produces a flatter probability
    distribution.

    Registered as a `Sampler` with name "top-k".
```